src/io/iohk/scalanet/peergroup/TLSPeerGroup.scala

```diff
@@ -0,0 +1,298 @@
+  * that are not instances of TCPPeerGroup.
+  *
+  * @param config bind address etc. See the companion object.
+  * @param codec a decco codec for reading writing messages to NIO ByteBuffer.
+  * @tparam M the message type.
+  */
+class TLSPeerGroup[M](val config: Config)(implicit codec: Codec[M]) extends TerminalPeerGroup[InetMultiAddress, M]() {
+
+  private val log = LoggerFactory.getLogger(getClass)
+
+  private[scalanet] class ServerChannelImpl[M](val nettyChannel: SocketChannel)(
+      implicit codec: Codec[M]
+  ) extends Channel[InetMultiAddress, M] {
+
+    private val log = LoggerFactory.getLogger(getClass)
+    private val messageSubject = ReplaySubject[M]()
+
+    private val sslServerCtx = SslContextBuilder.forServer(config.certChainPrivateKey, config.certChain: _*).build()
+
+    log.debug(
+      s"Creating server channel from ${nettyChannel.localAddress()} to ${nettyChannel.remoteAddress()} with channel id ${nettyChannel.id}"
+    )
+
+    nettyChannel
+      .pipeline()
+      .addLast("ssl", sslServerCtx.newHandler(nettyChannel.alloc())) //This needs to first
+      .addLast("frameDecoder", new LengthFieldBasedFrameDecoder(Int.MaxValue, 0, 4, 0, 4))
+      .addLast("frameEncoder", new LengthFieldPrepender(4))
+      .addLast(new MessageNotifier(messageSubject))
+
+    override val to: InetMultiAddress = InetMultiAddress(nettyChannel.remoteAddress())
+
+    override def sendMessage(message: M): Task[Unit] = {
+      toTask({
+        nettyChannel
+          .writeAndFlush(Unpooled.wrappedBuffer(codec.encode(message)))
+      })
+    }
+
+    override def in: Observable[M] = messageSubject
+
+    override def close(): Task[Unit] = {
+      messageSubject.onComplete()
+      toTask(nettyChannel.close())
+    }
+
+  }
+
+  private class ClientChannelImpl[M](inetSocketAddress: InetSocketAddress, clientBootstrap: Bootstrap)(
+      implicit codec: Codec[M]
+  ) extends Channel[InetMultiAddress, M] {
+
+    private val log = LoggerFactory.getLogger(getClass)
+
+    val to: InetMultiAddress = InetMultiAddress(inetSocketAddress)
+
+    private val activation = Promise[ChannelHandlerContext]()
+    private val activationF = activation.future
+
+    private val deactivation = Promise[Unit]()
+    private val deactivationF = deactivation.future
+
+    private val messageSubject = ReplaySubject[M]()
+
+    private val sslClientCtx = SslContextBuilder
+      .forClient()
+      .trustManager(config.trustStore: _*)
+      .build()
+
+    private val bootstrap: Bootstrap = clientBootstrap
+      .clone()
+      .handler(new ChannelInitializer[SocketChannel]() {
+        def initChannel(ch: SocketChannel): Unit = {
+          val pipeline = ch.pipeline()
+          val sslHandler = sslClientCtx.newHandler(ch.alloc())
+          pipeline
+            .addLast("ssl", sslHandler) //This needs to first
+            .addLast("frameEncoder", new LengthFieldPrepender(4))
```

Comment:
These lines are equivalent to what is being done on lines around `51`, create some constants and reuse them here.